### PR TITLE
fix #526

### LINKED ISSUE #526: Should add null check when put data into concurrenthashmap
The code here:
https://github.com/Microsoft/ApplicationInsights-Java/blob/v1.0.9/core/src/main/java/com/microsoft/applicationinsights/internal/util/MapUtil.java#L43

put data into a ConCurrentHashMap but does not have null check.

which will casue an NPE error if the `entry.getValue()` is null.

### PR DESCRIPTION
**Pull request by jdneo**
Fix #526  .

For significant contributions please make sure you have completed the following items:

- [X] Design discussion issue #
- [ ] Changes in public surface reviewed
- [X] CHANGELOG.md updated

### CODE REVIEW

**Comment by jdneo** [comment]
The `MapUtilTest` can pass locally, but I cannot run the whole peoject's test cases since I cannot build it on my side. Could please help verify it?

**Comment by dhaval24** [comment]
@jdneo I would really like to have the possibility for you to build the SDK locally otherwise it would inhibit possibility for you to have further meaningful contributions from you to our SDK which we really need and value a lot. I will look to set up everything on fresh VM and notedown the steps. I would first give it a shot with Visual Studio 2017 64 bit. Alternatively if you have linux VM you can build on that very successfully and this issue won't pop up.

**Comment by jdneo** [comment]
@dhaval24 That's great if there is a build step in detail. And maybe you can also write the steps into the `CONTRIBUTION.md` so that other people can also build this tool easily on their machine.

Another thing just curious, If this tool can build smoothly on Linux, why not setting up a travis CI server to run tests for each PR? I think it might be helpful to guard the master branch.

**Comment by dhaval24** [comment]
@jdneo we already have integration on VSTS for our Build. The current issue is that it only fires build for PRs by contributors who have are mentioned on this repository's contributor list. I.E doesn't fire for external contributions, we are looking to fix that issue soon. If you have ideas would be great. Our build could technically run all tests, drops artifacts on storage location which could be then picked by Azure Jenkins Java release job for signing and staging artifacts to maven. There it won't make sense to have a Travis-CI. 

Though there are few things we need to incorporate :  
1. Style check #490 
2. Code Coverage tools 
Any ideas for above two are welcomed

Yes I agree that we need to improve our contribution.md guide and infact there is a task also open on github for the same #489

**Comment by jdneo** [comment]
Ok, I got it.

**Comment by dhaval24** [comment]
@jdneo could you push a superfluous commit to this branch (simple comment to test out). I have tried to modify the build script to run CI on external PR's . would like to see if its working

**Comment by jdneo** [comment]
@dhaval24 seems it works

**Comment by dhaval24** [comment]
@jdneo seems to be working and tests are passing if you try to look at the details but apparently the overall CI Job still fails on the phase publish artifact drop with the following exception on the Console of build : Microsoft.VisualStudio.Services.Common.VssUnauthorizedException: TF400813: Resource not available for anonymous access. Client authentication required.
When it tries to publish the artifact to the folder. Don't know why this is happening.. Do you have any clue for this?

**Comment by jdneo** [comment]
have no idea on that 😢

**Comment by dhaval24** [comment]
@jdneo no worries please push back the change-log edit you did and we would try to figure out :-)

**Comment by jdneo** [comment]
Should I add the change log under v2.0.0-BETA? Or create a new title `v2.0.1` and add there?

**Comment by dhaval24** [comment]
v2.0.1

**Comment by dhaval24** [comment]
@littleaj @grlima please review this changes. I think these are good to go according to me. @littleaj we need to find a work around on one of the job failing in the build when it triggers on external PRs.

PS: I edited the build script to execute when external PR come.

File: core/src/main/java/com/microsoft/applicationinsights/internal/util/MapUtil.java [hunk 1]
```diff
@@ -33,14 +36,21 @@
...
         for (Map.Entry<String,Value> entry : source.entrySet()) {
             String key = entry.getKey();
             if (Strings.isNullOrEmpty(key)) {
                 continue;
             }
 
             if (!target.containsKey(key)) {
-                target.put(key,entry.getValue());
+                if (target instanceof ConcurrentHashMap && entry.getValue() == null) {
+                    InternalLogger.INSTANCE.trace("Skipping key '%s' while copying because of its mapping value is null", key);
```
Comment (littleaj) [inline core/src/main/java/com/microsoft/applicationinsights/internal/util/MapUtil.java]: Utility methods shouldn't log messages. remove this.

File: core/src/main/java/com/microsoft/applicationinsights/internal/util/MapUtil.java [hunk 2]
```diff
@@ -33,14 +36,21 @@
 public class MapUtil
 {
     public static <Value> void copy(Map<String, Value> source, Map<String, Value> target) {
+        if (source == null || target == null) {
+            return;
```
Comment (littleaj) [inline core/src/main/java/com/microsoft/applicationinsights/internal/util/MapUtil.java]: Let's use `Preconditions.checkArgument(source != null, "source must not be null")`. Similar for `target`.

**Comment by dhaval24** [comment]
@littleaj please take a look at these changes, @jdneo has addressed your feedback I think.

**Comment by dhaval24** [comment]
@littleaj if you think this is good to go please merge the changes in the master.

**Pull request merged**